### PR TITLE
Overhaul test_priorities.py

### PR DESCRIPTION
**Pull request by crusaderky**
- Make test_priorities.py deterministic
- Test scheduler-side sorting of unrannable tasks as well as worker-side sorting
- Blocks #6062

### CODE REVIEW

File: distributed/tests/test_priorities.py [hunk 1]
```diff
@@ -1,128 +1,131 @@
...
-        await wait(high)
-        assert all(s.processing.values())
-        assert s.tasks[low.key].state == "processing"
+
+    await wait(high)
+    assert all(s.processing.values())
+    assert s.tasks[low.key].state == "processing"
+    await ev.set()
+    await clog
+    await wait(low)
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_priorities.py:24]: > Call ensure_computing not at the end of the entire message batch but intermediately. I'm reasonably certain this test will fail with 

However, the original one should succeed with the changes proposed in #6062

**Comment by fjetter** [comment]
Is the stuff in here still flaky? do you need another review?

**Comment by crusaderky** [comment]
I think I found an important design problem.

Patch to #6077:
```patch
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -1831,6 +1831,7 @@ class Worker(ServerNode):
         annotations: dict | None = None,
         stimulus_id: str,
     ) -> None:
+        print(f"handle_compute_task({key}); {self.tasks=}")
         self.log.append((key, "compute-task", stimulus_id, time()))
         try:
             ts = self.tasks[key]
@@ -3372,6 +3373,8 @@ class Worker(ServerNode):
     def ensure_computing(self) -> None:
         if self.status in (Status.paused, Status.closing_gracefully):
             return
+        if self.ready and self.executing_count < self.nthreads:
+            print(f"ensure_computing: ready={sorted(self.ready)} {self.tasks=}")
         try:
             stimulus_id = f"ensure-computing-{time()}"
             while self.constrained and self.executing_count < self.nthreads:
```
Patch to this PR:
```patch
--- a/distributed/worker.py
+++ b/distributed/worker.py
@@ -1830,6 +1830,7 @@ class Worker(ServerNode):
         annotations: dict | None = None,
         stimulus_id: str,
     ) -> None:
+        print(f"handle_compute_task({key}); {self.tasks=}")
         self.log.append((key, "compute-task", stimulus_id, time()))
         try:
             ts = self.tasks[key]
@@ -3367,7 +3368,8 @@ class Worker(ServerNode):
     def _ensure_computing(self) -> RecsInstrs:
         if self.status in (Status.paused, Status.closing_gracefully):
             return {}, []
-
+        if self.ready and self.executing_count < self.nthreads:
+            print(f"ensure_computing: ready={sorted(self.ready)} {self.tasks=}")
         recs: Recs = {}
         while self.constrained and len(self._executing) < self.nthreads:
             key = self.constrained[0]
```

Output of test_priorities::test_compute():
#6077:
```
handle_compute_task(low); self.tasks={}
handle_compute_task(clog); self.tasks={'low': <TaskState 'low' ready>}
handle_compute_task(high); self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>}
ensure_computing: ready=[((-1, 1, 0, -2), 'high'), ((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' ready>}
ensure_computing: ready=[((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' memory>}
handle_compute_task(low); self.tasks={}
handle_compute_task(clog); self.tasks={'low': <TaskState 'low' ready>}
handle_compute_task(high); self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>}
ensure_computing: ready=[((-1, 1, 0, -2), 'high'), ((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' ready>}
ensure_computing: ready=[((0, 1, 0, -1), 'clog'), ((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' ready>, 'high': <TaskState 'high' memory>}
ensure_computing: ready=[((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' memory>, 'high': <TaskState 'high' memory>}
ensure_computing: ready=[((1, 1, 0, 0), 'low')] self.tasks={'low': <TaskState 'low' ready>, 'clog': <TaskState 'clog' memory>, 'high': <TaskState 'high' memory>}
```
This PR:
```
handle_compute_task(clog); self.tasks={}
ensure_computing: ready=[((0, 1, 0, 0), 'clog')] self.tasks={'clog': <TaskState 'clog' ready>}
handle_compute_task(high); self.tasks={'clog': <TaskState 'clog' executing>}
handle_compute_task(low); self.tasks={'clog': <TaskState 'clog' executing>, 'high': <TaskState 'high' ready>}
handle_compute_task(clog); self.tasks={}
ensure_computing: ready=[((0, 1, 0, 0), 'clog')] self.tasks={'clog': <TaskState 'clog' ready>}
handle_compute_task(high); self.tasks={'clog': <TaskState 'clog' executing>}
handle_compute_task(low); self.tasks={'clog': <TaskState 'clog' executing>, 'high': <TaskState 'high' ready>}
Dumped cluster state to test_cluster_dump/test_compute.yaml
ensure_computing: ready=[((-1, 1, 0, -1), 'high'), ((1, 1, 0, -2), 'low')] self.tasks={'clog': <TaskState 'clog' error>, 'high': <TaskState 'high' ready>, 'low': <TaskState 'low' ready>}
```

In both branches, the order of the first three handle_compute_task calls is random and changes at every run.
I verified that this matches the order of the calls to ``scheduler.py::_task_to_msg`` on the scheduler side.

In main, ensure_computing is not executed until all three have reached the worker. Hence, it finds a complete ``ready`` heap and can deterministically pick up the task with the lowest priority (high).

In this PR, ensure_computing is executed immediately after the first call to handle_compute (which sounds a lot more robust to me), but the obvious consequence is that the ready heap is incomplete and just contains a random task in it when this happens.

The correct solution IMHO is to ensure the scheduler invokes handle_compute_task in the correct order to begin with?

As a side note, I can't understand why there are more than 3 calls to handle_compute_task and why the task states are flip-flopping from memory back to ready.

**Comment by crusaderky** [comment]
Corollary: I have yet to prove it, but I believe that, *in main*, this means that as soon as you exceed the chunk size for batched sends then priority goes out of the window.

**Comment by crusaderky** [comment]
@fjetter this is good to be merged if you're happy with it

File: distributed/tests/test_priorities.py [hunk 2]
```diff
@@ -1,128 +1,166 @@
...
+    ys = [dinc(xs[i], dask_key_name=f"y{i}") for i in range(10)]
+    zs = [dinc(i, dask_key_name=f"z{i}") for i in range(10)]
+    ys = c.persist(ys)
+    zs = c.persist(zs)
+
+    while len(s.tasks) < 30:
+        await asyncio.sleep(0.01)
 
+    for fut in xs + ys + zs:
+        assert s.tasks[fut.key].priority[:2] == (0, 1)
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_priorities.py]: I rewrote the last two tests to verify the interleaving

File: distributed/tests/test_priorities.py [hunk 3]
```diff
@@ -1,128 +1,183 @@
...
+        await wait(ys + zs)
+        story = a.story("x0", "x1", "x2", "y0", "y1", "y2", "z0", "z1", "z2")
+
+    order = [ev[0] for ev in story if ev[1:3] == ("ready", "executing")]
+    # e.g. order = ['x0', 'z0', 'z1', 'z2', 'x1', 'x2', 'y0', 'y1', 'y2']
+    # There's an element of randomness, but there must be *some* interleaving of x and z
+    assert order[-3:] == ["y0", "y1", "y2"]
+    assert order.index("x0") < order.index("x1") < order.index("x2")
+    assert order.index("z0") < order.index("z1") < order.index("z2")
+    assert order.index("x0") < order.index("z2")
```
Comment (fjetter) [inline distributed/tests/test_priorities.py:162]: I honestly considered the previous version more easy to read and maintain. changing the 5 to a 9 in the `sum(t.state ...)` condition made the test pass for me 1000 times while the 5 indeed was flaky.

This version even fails for me locally.

**Review by fjetter — COMMENTED** [review]
The one concern I have now is that all tests are testing with a Worker that is spinning up after the task submission.
This takes a different code path than if workers are available during submission.

For instance, this entire file is now insensitive to whatever we do in update_graph. If I were to remove the sorting in https://github.com/dask/distributed/blob/a610f73ea2beea3715d08f6d4b70f930da0b76fd/distributed/scheduler.py#L4942 I would expect something in this file to break but it doesn't. However, I just checked, it is also not failing on main :(

File: distributed/tests/test_priorities.py [hunk 4]
```diff
@@ -1,128 +1,232 @@
...
+
+
+@asynccontextmanager
+async def block_worker(
+    c: Client,
+    s: Scheduler,
+    w: Worker,
+    pause: bool,
+    ntasks_on_scheduler: int | None = None,
+    ntasks_on_worker: int | None = None,
```
Comment (fjetter) [inline distributed/tests/test_priorities.py:26]: We're not using these anywhere, are we? Why expose this as part of the signature?

File: distributed/tests/test_priorities.py [hunk 5]
```diff
@@ -1,128 +1,232 @@
...
+):
+    """Make sure that no tasks submitted inside this context manager start running on
+    the worker until the context manager exits.
```
Comment (fjetter) [inline distributed/tests/test_priorities.py:29]: nit: This only works if there is one worker on the cluster and one thread on the worker.

**Review by fjetter — APPROVED** [review]
The test setup is much more complicated that I'd like it to be. However, the test cases themselves are easily understandable and the parametrization provides genuine value, so I'm OK with it

File: distributed/tests/test_priorities.py (hunk 4 above)
Comment (crusaderky) [reply distributed/tests/test_priorities.py:26]: we're using it in the last 2 tests

File: distributed/tests/test_priorities.py (hunk 5 above)
Comment (crusaderky) [reply distributed/tests/test_priorities.py:29]: clarified

**Pull request merged**